test/mip_test.c

```diff
@@ -65,204 +59,9 @@ static void test_statechange(void) {
-}
-
-static void test_http_fetch(void) {
-  // Setup interface
-  const char *iface = "tap0";             // Network iface
-  const char *mac = "00:00:01:02:03:78";  // MAC address
-  int fd = open("/dev/net/tun", O_RDWR);  // Open network interface
-
-  struct ifreq ifr;
-  memset(&ifr, 0, sizeof(ifr));
-  strncpy(ifr.ifr_name, iface, IFNAMSIZ);
-  ifr.ifr_flags = IFF_TAP | IFF_NO_PI;
-  if (ioctl(fd, TUNSETIFF, (void *) &ifr) < 0) {
-    MG_ERROR(("Failed to setup TAP interface: %s", ifr.ifr_name));
-    abort(); // return EXIT_FAILURE;
-  }
-  fcntl(fd, F_SETFL, fcntl(fd, F_GETFL, 0) | O_NONBLOCK);  // Non-blocking mode
-
-  MG_INFO(("Opened TAP interface: %s", iface));
-
-  // Events
-  struct mg_mgr mgr;  // Event manager
-  mg_mgr_init(&mgr);  // Initialise event manager
-
-  // MIP driver
-
-  // Zero init fields required (C/C++ style diverge)
-  #ifndef __cplusplus
-  struct mip_driver driver = {.tx = tap_tx, .up = tap_up, .rx = tap_rx};
-  // DHCP
-  // struct mip_if mif = {.use_dhcp = true, .driver = &driver, .driver_data = &fd};
-  // Static
-  // 192.168.32.2/24 gw 192.168.32.1
-  struct mip_if mif = {.use_dhcp = false, \
-  .ip=0x0220a8c0 , .mask=0x00ffffff, .gw=0x0120a8c0, \
-  .driver = &driver, .driver_data = &fd};
-
-  #else
-  struct mip_driver driver {};
-    driver.tx = tap_tx;
-    driver.up = tap_up;
-    driver.rx = tap_rx;
-  struct mip_if mif {};
-//  mif.use_dhcp = true;   // DHCP
-    mif.use_dhcp = false;  // Static IP
-    mif.ip = 0x0220a8c0;   // 192.168.32.2
-    mif.mask = 0x00ffffff; // 255.255.255.0
-    mif.gw = 0x0120a8c0;   // 192.168.32.1
-    mif.driver = &driver;
-    mif.driver_data = &fd;
-  #endif
-
-  sscanf(mac, "%hhx:%hhx:%hhx:%hhx:%hhx:%hhx", &mif.mac[0], &mif.mac[1], &mif.mac[2],
-         &mif.mac[3], &mif.mac[4], &mif.mac[5]);
-
-  mip_init(&mgr, &mif);
-  MG_INFO(("Init done, starting main loop"));
-
-  // Stack initialization, Network configuration (DHCP lease, ...)
-  {
-    if (mif.ip) MG_INFO(("MIF configuration: Static IP"));
-    else MG_INFO(("MIF configuration: DHCP"));
-    MG_INFO(("Opened TAP interface: %s", iface));
-    // ASSERT(!mif.ip); // Check we are set for DHCP
-    int pc = 500; // Timout on DHCP lease 500 ~ approx 5s (typical delay <1s)
-    while (((pc--)>0) /*& !mif.ip*/) {
-      mg_mgr_poll(&mgr, 100);
-      usleep(10000); // 10 ms
-    }
-    if (!mif.ip) printf("No ip assigned (DHCP lease may have failed).\n");
-    ASSERT(mif.ip); // We have an IP (lease or static)
-  }
-
-  // Simple HTTP fetch
-  {
-    char* http_feedback = (char*)"";
-    const bool ipv6 = 0;
-    if (ipv6) {
-      http_feedback = fetch (&mgr, "ipv6.google.com",\
-                                   "GET/ HTTP/1.0\r\nHost: ipv6.google.com\r\n\r\n");
-    } else {
-      http_feedback = fetch (&mgr, "http://cesanta.com",\
-                                   "GET //robots.txt HTTP/1.0\r\nHost: cesanta.com\r\n\r\n");
-    }
-
-    ASSERT(*http_feedback != '\0'); // Received HTTP response ?
```

Comment:
`http_feedback` is overwritten by the function results, and it may be NULL.
I suggest:
```c
    char* http_feedback;
    const bool ipv6 = 0;
    if (ipv6) {
      http_feedback = fetch (&mgr, "ipv6.google.com",\
                                   "GET/ HTTP/1.0\r\nHost: ipv6.google.com\r\n\r\n");
    } else {
      http_feedback = fetch (&mgr, "http://cesanta.com",\
                                   "GET //robots.txt HTTP/1.0\r\nHost: cesanta.com\r\n\r\n");
    }

    ASSERT(http_feedback != NULL && *http_feedback != '\0'); // Received HTTP response ?
```

Reply:
Ok done this one.
Indeed null pointer and void response both needed.